frontend/src/components/widget/WidgetRenderer.js

```diff
@@ -0,0 +1,629 @@
+import DevicesWidget from './Devices/DevicesWidget';
+import Image from './Image';
+import Labels from './Labels';
+import Link from './Link';
+import CharacterLimitInfo from './CharacterLimitInfo';
+import List from './List/List';
+import Lookup from './Lookup/Lookup';
+
+class WidgetRenderer extends PureComponent {
+  constructor(props) {
+    super(props);
+
+    this.getClassNames = getClassNames.bind(this);
+    this.generateMomentObj = generateMomentObj.bind(this);
+  }
+
+  render() {
+    const {
+      handleChange,
+      updated,
+      isModal,
+      filterWidget,
+      filterId,
+      id,
+      range,
+      onHide,
+      handleBackdropLock,
+      subentity,
+      widgetType,
+      subentityId,
+      dropdownOpenCallback,
+      autoFocus,
+      fullScreen,
+      fields,
+      windowType,
+      dataId,
+      type,
+      widgetData,
+      rowId,
+      tabId,
+      docId,
+      activeTab,
+      icon,
+      gridAlign,
+      entity,
+      onShow,
+      caption,
+      viewId,
+      listenOnKeys,
+      listenOnKeysFalse,
+      closeTableField,
+      handleZoomInto,
+      attribute,
+      allowShowPassword,
+      onBlurWidget,
+      isOpenDatePicker,
+      dateFormat,
+      initialFocus,
+      timeZone,
+      fieldName,
+      maxLength,
+      updateHeight,
+      rowIndex,
+      onClickOutside,
+      emptyText,
+      forceFullWidth,
+      forceHeight,
+
+      // from `renderWidget`
+      isMultiselect,
+      widgetField,
+      widgetProperties,
+      showErrorBorder,
+      isFocused,
+      charsTyped,
+      readonly,
+      onPatch,
+      onListFocus,
+      onBlurWithParams,
+      onSetWidgetType,
+      onHandleProcess,
+      forwardedRef,
+    } = this.props;
+    const { tabIndex, onFocus, onBlur } = widgetProperties;
+    const widgetValue = get(widgetProperties, ['value'], null);
+    widgetProperties.ref = forwardedRef;
+
+    const selectedValue = widgetData[0].value
+      ? widgetData[0].value
+      : widgetData[0].defaultValue;
+
+    const listAndLookupsProps = {
+      dataId,
+      attribute,
+      entity,
+      subentity,
+      subentityId,
+      windowType,
+      readonly,
+      updated,
+      filterWidget,
+      filterId,
+      tabId,
+      rowId,
+      tabIndex,
+      viewId,
+      autoFocus,
+      onFocus: onListFocus,
+      onBlur: onBlurWithParams,
+      align: gridAlign,
+      mandatory: widgetData[0].mandatory,
+      parameterName: fields[0].parameterName,
+      validStatus: widgetData[0].validStatus,
+      onChange: onPatch,
+    };
+    const dateProps = {
+      field: widgetField,
+      key: 1,
+      handleChange,
+      handleBackdropLock,
+      inputProps: {
+        placeholder: fields[0].emptyText,
+        disabled: readonly,
+        tabIndex: tabIndex,
+      },
+      onChange: (date) => handleChange(widgetField, date),
+    };
+    const dateRangeProps = {
+      mandatory: widgetData[0].mandatory,
```

Comment:
would get those values before {mandatory,  validStatus } = widgetData[0];  also would add some safety checks

Reply:
`WidgetWrapper` returns `widgetData` as `[{}]` so this will just be falsy. One one hand having checks makes sense, but in our case we'd have to add them everywhere, where we're checking directly a property in widgetData object.